File: 0706-design-hashmap/solution.ts
// Solution: Design HashMap
// Runtime: 26 ms | Memory: 68.6 MB

class MyHashMap {
  private size: number;
  private buckets: [number, number][][]; // [0 , [[1,2],[2,4]]]

  constructor() {
    this.size = 1000;
    this.buckets = new Array(this.size).fill(null).map(() => []);
  }

  private hash(key: number): number {
    return key % this.size;
  }

  put(key: number, value: number): void {
    const index = this.hash(key);
    const bucket = this.buckets[index];

    for (let i = 0; i < bucket!.length; i++) {
      if (bucket[i][0] === key) {
        bucket[i][1] = value;
        return;
      }
    }

    bucket?.push([key, value]);
  }

  get(key: number): number {
    const index = this.hash(key);
    const bucket = this.buckets[index];

    for (let i = 0; i < bucket!.length; i++) {
      if (bucket[i][0] === key) {
        return bucket[i][1];
      }
    }

    return -1;
  }

  remove(key: number): void {
    const index = this.hash(key);
    const bucket = this.buckets[index];

    for (let i = 0; i < bucket!.length; i++) {
      if (bucket[i][0] === key) {
        bucket?.splice(i, 1);
      }
    }
  }
}
/**
 * Your MyHashMap object will be instantiated and called as such:
 * var obj = new MyHashMap()
 * obj.put(key,value)
 * var param_2 = obj.get(key)
 * obj.remove(key)
 */

File: 0706-design-hashmap/solution.cpp
// Solution: Design HashMap
// Runtime: 13 ms | Memory: 59.8 MB

#include <vector>
using namespace std;

class MyHashMap {
private:
    int size;
    vector<vector<pair<int, int>>> buckets;

    int hash(int key) {
        return key % size;
    }

public:
    MyHashMap() {
        size = 1000;
        buckets.resize(size);
    }

    void put(int key, int value) {
        int index = hash(key);
        auto& bucket = buckets[index];

        for (int i = 0; i < bucket.size(); i++) {
            if (bucket[i].first == key) {
                bucket[i].second = value;
                return;
            }
        }

        bucket.push_back({key, value});
    }

    int get(int key) {
        int index = hash(key);
        auto& bucket = buckets[index];

        for (int i = 0; i < bucket.size(); i++) {
            if (bucket[i].first == key) {
                return bucket[i].second;
            }
        }

        return -1;
    }

    void remove(int key) {
        int index = hash(key);
        auto& bucket = buckets[index];

        for (int i = 0; i < bucket.size(); i++) {
            if (bucket[i].first == key) {
                bucket.erase(bucket.begin() + i);
                return;
            }
        }
    }
};

/**
 * Your MyHashMap object will be instantiated and called as such:
 * MyHashMap* obj = new MyHashMap();
 * obj->put(key, value);
 * int param_2 = obj->get(key);
 * obj->remove(key);
 */